server/ServerFSM.cpp

```diff
@@ -274,7 +274,7 @@ void ServerFSM::HandleNonLobbyDisconnection(const Disconnection& d) {
         }
         // add "player left game" message
         boost::posix_time::ptime timestamp = boost::posix_time::second_clock::universal_time();
-        std::string data = player_connection->PlayerName() + " [[" + UserStringNop("PLAYER_LEFT_GAME") + "]]";
+        std::string data = (boost::format("%s [[PLAYER_LEFT_GAME]]") % player_connection->PlayerName()).str();
```

Comment:
This is no different from using `+`.

The point is that the string composition should be done client-side so the stringtable of the client can determine how and where to put the parameter into the string.

Reply:
Not all languages have Subject-Verb-Object ordering, so you shouldn't assume the player name will appear first in the string. If the substitution is done on the client, and has a playername parameter, the stringtable can specify where the substituted name should appear in the displayed text.

Reply:
Does `[[STRINGTABLE_ENTRY]]` have arguments support so it could be written like `[[PLAYER_LEFT_GAME,Player name]]`?

Reply:
No, see here:

https://github.com/freeorion/freeorion/blob/master/UI/ChatWnd.cpp#L37

It's a bit hard to follow, but it scans through the input chat message, copying everything to the filtered message until it sees `[[`, then it searches for a stringtable entry that matches what comes next until the following `]]`. It then copies the stringtable's value instead of the `[[` and `]]` and the stuff between them.

No format call occurs, so no formatting gets done like for `VarText`, and no recursive macro substitution is done, like when loading the stringtable or scripts and processing macros.

Reply:
Ok, I'll postpone this PR until some stringtable arguments would be implemented.

Reply:
https://github.com/freeorion/freeorion/commit/783fa637b5733a2730aa69e18ff441daa4aaad1c should support client-side string or stringtable entry substitutions.

Reply:
Thanks, updated PR according to this change.